src/irr/video/COpenGLSpecializedShader.cpp

```diff
@@ -114,10 +114,104 @@ COpenGLSpecializedShader::COpenGLSpecializedShader(const video::IVideoDriver* _d
 
     GLchar logbuf[1u<<12]; //4k
     driver->extGlGetProgramInfoLog(m_GLname, sizeof(logbuf), nullptr, logbuf);
-    os::Printer::log(logbuf, ELL_ERROR);
+    if (logbuf[0])
+        os::Printer::log(logbuf, ELL_ERROR);
 
-    // TODO:
-    // what should be interface for setting push_constants (regular uniform on GL backend) for now?
+    m_introspectionData = core::smart_refctd_ptr<asset::CIntrospectionData>(_introspection);
+}
+
+void COpenGLSpecializedShader::setUniformsImitatingPushConstants(const uint8_t* _pcData)
+{
+    if (m_uniformsList.empty())
+        buildUniformsList();
+
+    using gl = COpenGLExtensionHandler;
+    for (const SUniform& u : m_uniformsList)
+    {
+        const SMember& m = u.m;
+        auto is_mtx = [&m] { return (m.mtxRowCnt>1u && m.mtxColCnt>1u); };
+        auto is_single_or_vec = [&m] { return (m.mtxRowCnt>=1u && m.mtxColCnt==1u); };
+
+        if (is_mtx() && m.type==asset::EGVT_F32) {
+            core::vector<GLfloat> matrix_data(m.mtxRowCnt*m.mtxColCnt*m.count);
+            for (uint32_t i = 0u; i < m.count; ++i)
+            {
+                for (uint32_t c = 0u; c < m.mtxColCnt; ++c) {
+                    const GLfloat* col = reinterpret_cast<const GLfloat*>(_pcData + m.offset + i*m.arrayStride + c*m.mtxStride);
```

Comment:
find out if the offsets into the push_constant range need to be somehow aligned for the types (on the GLSL side) and put some assertions here

Reply:
_pcData is supposed to be already offset'd range of those 128 bytes, but not sure if I understand u correctly

Reply:
Im pretty sure that the `col` pointer needs to have ome sort of alignment before you pass it to setUniform, i.e. 4 bytes or something

Reply:
@Crisspl ?

Reply:
i havent seen anywhere such requirement

Reply:
id assume that an opengl function taking a type T will fuck up if pointer is not aligned to alingof(T) would be nice to at least have an assert

Reply:
please do an alignment check.

Reply:
resolved